File: internal/content/content.go
package content

import (
	"fmt"
	"html/template"
	"math"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"
	"unicode/utf8"

	"github.com/93percentai/sitemaker/internal/config"
	"github.com/93percentai/sitemaker/internal/directive"
)

type Page struct {
	SourcePath  string
	RawContent  []byte
	Body        []byte
	Directives  directive.Directives
	HTMLContent template.HTML
	TOCContent  template.HTML

	Title       string
	Slug        string
	URL         string
	Date        time.Time
	Tags        []string
	ReadingTime int
	IsBlogPost  bool
	Draft       bool

	Prev *Page
	Next *Page
}

type Site struct {
	Config   config.Config
	Pages    []*Page
	Posts    []*Page
	TagIndex map[string][]*Page
}

var datePrefix = regexp.MustCompile(`^(\d{4}-\d{2}-\d{2})-(.+)$`)

func LoadSite(cfg config.Config) (*Site, error) {
	site := &Site{
		Config:   cfg,
		TagIndex: make(map[string][]*Page),
	}

	contentDir := cfg.Build.ContentDir
	err := filepath.Walk(contentDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() {
			return nil
		}
		if filepath.Ext(path) != ".md" {
			return nil
		}

		page, err := loadPage(path, contentDir, cfg)
		if err != nil {
			return fmt.Errorf("loading %s: %w", path, err)
		}

		site.Pages = append(site.Pages, page)
		return nil
	})

	if err != nil {
		return nil, err
	}

	return site, nil
}

func loadPage(path, contentDir string, cfg config.Config) (*Page, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	relPath, _ := filepath.Rel(contentDir, path)
	relPath = filepath.ToSlash(relPath)

	directives, body := directive.Parse(data)

	page := &Page{
		SourcePath: relPath,
		RawContent: data,
		Body:       body,
		Directives: directives,
		Draft:      directives.Draft,
		Tags:       directives.Tags,
	}

	page.IsBlogPost = isPost(relPath, cfg)
	page.resolveTitle()
	page.resolveSlug(cfg)
	page.resolveDate(cfg)
	page.resolveURL(cfg)
	page.ReadingTime = estimateReadingTime(body)

	return page, nil
}

func isPost(relPath string, cfg config.Config) bool {
	return strings.HasPrefix(relPath, cfg.Blog.PostsDir+"/")
}

func (p *Page) resolveTitle() {
	if p.Directives.Title != "" {
		p.Title = p.Directives.Title
		return
	}
	for _, line := range strings.Split(string(p.Body), "\n") {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "# ") {
			p.Title = strings.TrimPrefix(line, "# ")
			return
		}
	}
	base := filepath.Base(p.SourcePath)
	p.Title = strings.TrimSuffix(base, filepath.Ext(base))
}

func (p *Page) resolveSlug(cfg config.Config) {
	if p.Directives.Slug != "" {
		p.Slug = p.Directives.Slug
		return
	}

	base := strings.TrimSuffix(filepath.Base(p.SourcePath), ".md")

	if p.IsBlogPost && cfg.URLs.StripDate {
		if m := datePrefix.FindStringSubmatch(base); m != nil {
			base = m[2]
		}
	}

	p.Slug = base
}

func (p *Page) resolveDate(cfg config.Config) {
	if p.Directives.Date != "" {
		t, err := time.Parse(cfg.Blog.DateFormat, p.Directives.Date)
		if err == nil {
			p.Date = t
			return
		}
	}

	base := strings.TrimSuffix(filepath.Base(p.SourcePath), ".md")
	if m := datePrefix.FindStringSubmatch(base); m != nil {
		t, err := time.Parse("2006-01-02", m[1])
		if err == nil {
			p.Date = t
		}
	}
}

func (p *Page) resolveURL(cfg config.Config) {
	dir := filepath.Dir(p.SourcePath)
	if dir == "." {
		dir = ""
	}

	slug := p.Slug

	if slug == "index" {
		if dir == "" {
			p.URL = "/"
		} else {
			p.URL = "/" + dir + "/"
		}
		return
	}

	var urlPath string
	if dir != "" {
		urlPath = dir + "/" + slug
	} else {
		urlPath = slug
	}

	switch cfg.URLs.Style {
	case "directory":
		p.URL = "/" + urlPath + "/"
	default:
		p.URL = "/" + urlPath + ".html"
	}
}

func (p *Page) OutputPath(cfg config.Config) string {
	if p.Slug == "index" {
		dir := filepath.Dir(p.SourcePath)
		if dir == "." {
			return "index.html"
		}
		return filepath.Join(dir, "index.html")
	}

	dir := filepath.Dir(p.SourcePath)
	if dir == "." {
		dir = ""
	}

	switch cfg.URLs.Style {
	case "directory":
		return filepath.Join(dir, p.Slug, "index.html")
	default:
		return filepath.Join(dir, p.Slug+".html")
	}
}

func (p *Page) HasHeader() bool   { return p.Directives.HasComponent("H") }
func (p *Page) HasFooter() bool   { return p.Directives.HasComponent("F") }
func (p *Page) HasTOC() bool      { return p.Directives.HasComponent("TOC") }
func (p *Page) HasSidebar() bool  { return p.Directives.HasComponent("SB") }

func (p *Page) CSSAssets() []string {
	var css []string
	for _, a := range p.Directives.AssetFiles() {
		if strings.HasSuffix(a, ".css") {
			css = append(css, a)
		}
	}
	return css
}

func (p *Page) JSAssets() []string {
	var js []string
	for _, a := range p.Directives.AssetFiles() {
		if strings.HasSuffix(a, ".js") {
			js = append(js, a)
		}
	}
	return js
}

func estimateReadingTime(content []byte) int {
	words := utf8.RuneCount(content) / 5
	minutes := int(math.Ceil(float64(words) / 200.0))
	if minutes < 1 {
		minutes = 1
	}
	return minutes
}

File: src/content.ts
import fs from "node:fs";
import path from "node:path";
import { Config } from "./config.js";
import { Directives, parseDirectives, hasComponent, assetFiles } from "./directive.js";

export interface Page {
  sourcePath: string;
  rawContent: string;
  body: string;
  directives: Directives;
  htmlContent: string;
  tocContent: string;
  title: string;
  slug: string;
  url: string;
  date: Date | null;
  dateStr: string;
  tags: string[];
  readingTime: number;
  isBlogPost: boolean;
  draft: boolean;
  hasHeader: boolean;
  hasFooter: boolean;
  hasTOC: boolean;
  hasSidebar: boolean;
  cssAssets: string[];
  jsAssets: string[];
  prev: { title: string; url: string } | null;
  next: { title: string; url: string } | null;
}

export interface Site {
  config: Config;
  pages: Page[];
  posts: Page[];
  tagIndex: Record<string, Page[]>;
}

const DATE_PREFIX_RE = /^(\d{4}-\d{2}-\d{2})-(.+)$/;

export function loadSite(cfg: Config): Site {
  const site: Site = {
    config: cfg,
    pages: [],
    posts: [],
    tagIndex: {},
  };

  walkDir(cfg.build.content_dir, cfg.build.content_dir, site, cfg);
  return site;
}

function walkDir(dir: string, contentDir: string, site: Site, cfg: Config): void {
  if (!fs.existsSync(dir)) return;

  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      walkDir(fullPath, contentDir, site, cfg);
    } else if (entry.name.endsWith(".md")) {
      const page = loadPage(fullPath, contentDir, cfg);
      site.pages.push(page);
    }
  }
}

function loadPage(filePath: string, contentDir: string, cfg: Config): Page {
  const raw = fs.readFileSync(filePath, "utf-8");
  const relPath = path.relative(contentDir, filePath).replace(/\\/g, "/");
  const { directives, body } = parseDirectives(raw);

  const isBlogPost = relPath.startsWith(cfg.blog.posts_dir + "/");
  const title = resolveTitle(directives, body, relPath);
  const slug = resolveSlug(directives, relPath, isBlogPost, cfg);
  const date = resolveDate(directives, relPath);
  const url = resolveURL(relPath, slug, cfg);

  const assets = assetFiles(directives);

  return {
    sourcePath: relPath,
    rawContent: raw,
    body,
    directives,
    htmlContent: "",
    tocContent: "",
    title,
    slug,
    url,
    date,
    dateStr: date ? formatDate(date) : "",
    tags: directives.tags,
    readingTime: estimateReadingTime(body),
    isBlogPost,
    draft: directives.draft,
    hasHeader: hasComponent(directives, "H"),
    hasFooter: hasComponent(directives, "F"),
    hasTOC: hasComponent(directives, "TOC"),
    hasSidebar: hasComponent(directives, "SB"),
    cssAssets: assets.filter((a) => a.endsWith(".css")),
    jsAssets: assets.filter((a) => a.endsWith(".js")),
    prev: null,
    next: null,
  };
}

function resolveTitle(d: Directives, body: string, relPath: string): string {
  if (d.title) return d.title;
  const match = body.match(/^#\s+(.+)$/m);
  if (match) return match[1];
  const base = path.basename(relPath, ".md");
  return base;
}

function resolveSlug(
  d: Directives,
  relPath: string,
  isBlogPost: boolean,
  cfg: Config
): string {
  if (d.slug) return d.slug;
  let base = path.basename(relPath, ".md");
  if (isBlogPost && cfg.urls.strip_date) {
    const m = base.match(DATE_PREFIX_RE);
    if (m) base = m[2];
  }
  return base;
}

function resolveDate(d: Directives, relPath: string): Date | null {
  if (d.date) {
    const t = new Date(d.date + "T00:00:00Z");
    if (!isNaN(t.getTime())) return t;
  }
  const base = path.basename(relPath, ".md");
  const m = base.match(DATE_PREFIX_RE);
  if (m) {
    const t = new Date(m[1] + "T00:00:00Z");
    if (!isNaN(t.getTime())) return t;
  }
  return null;
}

function resolveURL(relPath: string, slug: string, cfg: Config): string {
  const dir = path.dirname(relPath);
  const dirPrefix = dir === "." ? "" : dir;

  if (slug === "index") {
    return dirPrefix === "" ? "/" : `/${dirPrefix}/`;
  }

  const urlPath = dirPrefix ? `${dirPrefix}/${slug}` : slug;

  if (cfg.urls.style === "directory") {
    return `/${urlPath}/`;
  }
  return `/${urlPath}.html`;
}

export function outputPath(page: Page, cfg: Config): string {
  if (page.slug === "index") {
    const dir = path.dirname(page.sourcePath);
    if (dir === ".") return "index.html";
    return path.join(dir, "index.html");
  }

  const dir = path.dirname(page.sourcePath);
  const dirPrefix = dir === "." ? "" : dir;

  if (cfg.urls.style === "directory") {
    return path.join(dirPrefix, page.slug, "index.html");
  }
  return path.join(dirPrefix, page.slug + ".html");
}

export function classifyPages(site: Site): void {
  site.posts = site.pages
    .filter((p) => p.isBlogPost && !p.draft)
    .sort((a, b) => {
      const da = a.date?.getTime() ?? 0;
      const db = b.date?.getTime() ?? 0;
      return db - da;
    });
}

export function buildTagIndex(site: Site): void {
  site.tagIndex = {};
  for (const p of site.posts) {
    for (const tag of p.tags) {
      if (!site.tagIndex[tag]) site.tagIndex[tag] = [];
      site.tagIndex[tag].push(p);
    }
  }
}

export function linkPrevNext(site: Site): void {
  for (let i = 0; i < site.posts.length; i++) {
    if (i > 0) {
      site.posts[i].next = {
        title: site.posts[i - 1].title,
        url: site.posts[i - 1].url,
      };
    }
    if (i < site.posts.length - 1) {
      site.posts[i].prev = {
        title: site.posts[i + 1].title,
        url: site.posts[i + 1].url,
      };
    }
  }
}

function estimateReadingTime(content: string): number {
  const words = content.split(/\s+/).length;
  return Math.max(1, Math.ceil(words / 200));
}

function formatDate(d: Date): string {
  return d.toISOString().slice(0, 10);
}
